hack/cleanup-ag.sh

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env bash
+#
+# Copyright 2023 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires: jq, kubectl, cmk (get https://github.com/apache/cloudstack-cloudmonkey/releases/tag/6.4.0-rc1 or later)
+#
+# About: this tool helps to remove CloudStack affinity groups from CAPC
+# management cluster which are not assigned to any instances.
+#
+# Usage and help:
+# chmod +x cleanup-ag.sh
+# ./cleanup-ag.sh -h
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+export DRY_RUN=false
+export VERBOSE=false
+export KUBECONFIG=$HOME/.kube/config
+
+debug() {
+  if [[ "$VERBOSE" == "true" ]]; then
+    echo $@
+  fi
+}
+
+get_affinity_groups() {
+  kubectl get cloudstackaffinitygroups -o json -A | jq -r '.items[].metadata.name'
+}
+
+get_cluster() {
+  affinityGroup=$1
+  kubectl get cloudstackaffinitygroup $affinityGroup -o json | jq -r '.metadata.labels."cluster.x-k8s.io/cluster-name"'
+}
+
+get_cluster_credentials() {
+  cluster=$1
+  kubectl get cloudstackcluster $cluster -o json | jq -r '.spec.failureDomains[].acsEndpoint.name' | uniq
+}
+
+setup_acs_credentials() {
+  credential=$1
+  export CS_URL=$(kubectl get secret $credential -o json | jq -r '.data."api-url"' | base64 -D)
+  export CS_APIKEY=$(kubectl get secret $credential -o json | jq -r '.data."api-key"' | base64 -D)
+  export CS_SECRETKEY=$(kubectl get secret $credential -o json | jq -r '.data."secret-key"' | base64 -D)
+  debug "Using CloudStack Control Plane URL: $CS_URL and CloudStack Account: $(run_cmk list users | jq -r '.user[] | .account + " and User: " + .username')"
+}
+
+run_cmk() {
+  cmk -u $CS_URL -k $CS_APIKEY -s $CS_SECRETKEY -o json $@
+}
+
+main() {
+  for ag in $(get_affinity_groups); do
+    echo "Checking CloudStack Affinity Group: $ag"
+    cluster=$(get_cluster $ag)
+    for credential in $(get_cluster_credentials $cluster); do
+      setup_acs_credentials $credential
+      CS_AG_ID=$(kubectl get cloudstackaffinitygroup $ag -o json | jq -r '.spec.id')
```

Comment:
Specify namespace or use -A to get the id here.

Reply:
Same as above:
```
 > kubectl get cloudstackaffinitygroup antiaffinity-capc-cluster-control-plane-746e19e9-ea06-4ca3-b9b1-ebeb4753e6c7-failure-domain-1 -A
error: a resource cannot be retrieved by name across all namespaces
```